Question:
- How many times do people view the item before purchase.
- what time and what day have the highest traffic of people viewing products
- what date have the most sales and least sales.
- How loyal are people to a brand. (How often the same individual repurchases an items from the same brand.)
- When the person purchased the item, what other item did the person also purchase at the same time? (using the association rules)

In [10]:
import numpy as np 
import pandas as pd 
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('/Users/krits/Documents/GitHub/ICCS361Project1/DATA/2019-Oct.csv')
df

event_time event_type  product_id          category_id  \
0         2019-10-01 00:00:00 UTC       view    44600062  2103807459595387724   
1         2019-10-01 00:00:00 UTC       view     3900821  2053013552326770905   
2         2019-10-01 00:00:01 UTC       view    17200506  2053013559792632471   
3         2019-10-01 00:00:01 UTC       view     1307067  2053013558920217191   
4         2019-10-01 00:00:04 UTC       view     1004237  2053013555631882655   
...                           ...        ...         ...                  ...   
42448759  2019-10-31 23:59:58 UTC       view     2300275  2053013560530830019   
42448760  2019-10-31 23:59:58 UTC       view    10800172  2053013554994348409   
42448761  2019-10-31 23:59:58 UTC       view     5701038  2053013553970938175   
42448762  2019-10-31 23:59:59 UTC       view    21407424  2053013561579406073   
42448763  2019-10-31 23:59:59 UTC       view    13300120  2053013557166998015   

                                category_code      brand    price    user_id  \
0                                         NaN   shiseido    35.79  541312140   
1         appliances.environment.water_heater       aqua    33.20  554748717   
2                  furniture.living_room.sofa        NaN   543.10  519107250   
3                          computers.notebook     lenovo   251.74  550050854   
4                      electronics.smartphone      apple  1081.98  535871217   
...                                       ...        ...      ...        ...   
42448759             electronics.camera.video      gopro   527.40  537931532   
42448760                                  NaN    redmond    61.75  527322328   
42448761              auto.accessories.player    kenwood   128.70  566280422   
42448762                   electronics.clocks     tissot   689.85  513118352   
42448763                                  NaN  swisshome   155.73  525266378   

                                  user_session  
0         72d76fde-8bb3-4e00-8c23-a032dfed738c  
1         9333dfbd-b87a-4708-9857-6336556b0fcc  
2         566511c2-e2e3-422b-b695-cf8e6e792ca8  
3         7c90fc70-0e80-4590-96f3-13c02c18c713  
4         c6bd7419-2748-4c56-95b4-8cec9ff8b80d  
...                                        ...  
42448759  22c57267-da98-4f28-9a9c-18bb5b385193  
42448760  5054190a-46cb-4211-a8f1-16fc1a060ed8  
42448761  05b6c62b-992f-4e8e-91f7-961bcb4719cd  
42448762  4c14bf2a-2820-4504-929d-046356a5a204  
42448763  6e57d2d7-6022-46e6-81d6-fa77f14cefd8  

[42448764 rows x 9 columns]

In [12]:
df.head(10)


event_time event_type  product_id          category_id  \
0 2019-10-01 00:00:00+00:00       view    44600062  2103807459595387724   
1 2019-10-01 00:00:00+00:00       view     3900821  2053013552326770905   
2 2019-10-01 00:00:01+00:00       view    17200506  2053013559792632471   
3 2019-10-01 00:00:01+00:00       view     1307067  2053013558920217191   
4 2019-10-01 00:00:04+00:00       view     1004237  2053013555631882655   
5 2019-10-01 00:00:05+00:00       view     1480613  2053013561092866779   
6 2019-10-01 00:00:08+00:00       view    17300353  2053013553853497655   
7 2019-10-01 00:00:08+00:00       view    31500053  2053013558031024687   
8 2019-10-01 00:00:10+00:00       view    28719074  2053013565480109009   
9 2019-10-01 00:00:11+00:00       view     1004545  2053013555631882655   

                         category_code     brand    price    user_id  \
0                                  NaN  shiseido    35.79  541312140   
1  appliances.environment.water_heater      aqua    33.20  554748717   
2           furniture.living_room.sofa       NaN   543.10  519107250   
3                   computers.notebook    lenovo   251.74  550050854   
4               electronics.smartphone     apple  1081.98  535871217   
5                    computers.desktop    pulser   908.62  512742880   
6                                  NaN     creed   380.96  555447699   
7                                  NaN  luminarc    41.16  550978835   
8                   apparel.shoes.keds     baden   102.71  520571932   
9               electronics.smartphone    huawei   566.01  537918940   

                           user_session  
0  72d76fde-8bb3-4e00-8c23-a032dfed738c  
1  9333dfbd-b87a-4708-9857-6336556b0fcc  
2  566511c2-e2e3-422b-b695-cf8e6e792ca8  
3  7c90fc70-0e80-4590-96f3-13c02c18c713  
4  c6bd7419-2748-4c56-95b4-8cec9ff8b80d  
5  0d0d91c2-c9c2-4e81-90a5-86594dec0db9  
6  4fe811e9-91de-46da-90c3-bbd87ed3a65d  
7  6280d577-25c8-4147-99a7-abc6048498d6  
8  ac1cd4e5-a3ce-4224-a2d7-ff660a105880  
9  406c46ed-90a4-4787-a43b-59a410c1a5fb

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42448764 entries, 0 to 42448763
Data columns (total 9 columns):
 #   Column         Dtype              
---  ------         -----              
 0   event_time     datetime64[ns, UTC]
 1   event_type     object             
 2   product_id     int64              
 3   category_id    int64              
 4   category_code  object             
 5   brand          object             
 6   price          float64            
 7   user_id        int64              
 8   user_session   object             
dtypes: datetime64[ns, UTC](1), float64(1), int64(3), object(4)
memory usage: 2.8+ GB


In [16]:
df.describe()

product_id   category_id         price       user_id
count  4.244876e+07  4.244876e+07  4.244876e+07  4.244876e+07
mean   1.054993e+07  2.057404e+18  2.903237e+02  5.335371e+08
std    1.188191e+07  1.843926e+16  3.582692e+02  1.852374e+07
min    1.000978e+06  2.053014e+18  0.000000e+00  3.386938e+07
25%    1.005157e+06  2.053014e+18  6.598000e+01  5.159043e+08
50%    5.000470e+06  2.053014e+18  1.629300e+02  5.296965e+08
75%    1.600030e+07  2.053014e+18  3.585700e+02  5.515788e+08
max    6.050001e+07  2.175420e+18  2.574070e+03  5.662809e+08

In [17]:
df.count()

event_time       42448764
event_type       42448764
product_id       42448764
category_id      42448764
category_code    28933155
brand            36331684
price            42448764
user_id          42448764
user_session     42448762
dtype: int64

In [19]:
df.isnull().sum()

event_time              0
event_type              0
product_id              0
category_id             0
category_code    13515609
brand             6117080
price                   0
user_id                 0
user_session            2
dtype: int64

In [23]:
print(df['event_type'].nunique())
print(df['event_type'].unique())

3
['view' 'purchase' 'cart']


In [29]:
print(df['product_id'].nunique())
print(df['category_id'].nunique())
print(df['category_code'].nunique())
print(df['brand'].nunique())
print(df['user_id'].nunique())

166794
624
126
3444
3022290


In [30]:
event_type_percentages = df['event_type'].value_counts(normalize=True) * 100
print(event_type_percentages)

event_type
view        96.067341
cart         2.182669
purchase     1.749990
Name: proportion, dtype: float64


In [31]:
min_spending = df['price'].min()
max_spending = df['price'].max()
avg_spending = df['price'].mean()

print(f"Minimum spending: ${min_spending:.2f}")
print(f"Maximum spending: ${max_spending:.2f}")
print(f"Average spending: ${avg_spending:.2f}")

Minimum spending: $0.00
Maximum spending: $2574.07
Average spending: $290.32


Summary of DATA:

| Property    | Description |   Type   |   Note  |
| --------    | -------     | -------- | ------- |
| event_time  | Time when event happened at (in UTC).                                                                                                                          | datetime64  | total of 42448764 time stamp   |
| event_type  | Only one kind of event: purchase.                                                                                                                              | object |  3 event type in total view = 96%, cart = 2.2% and purchase = 1.7%   |
| product_id  | ID of a product                                                                                                                                                | int64    | 166794 unique product   |
| category_id | Product's category ID                                                                                                                                          | int64    | 624 unique category ID   |
|category_code| Product's category taxonomy (code name) if it was possible to make it. Usually present for meaningful categories and skipped for different kinds of accessories| object    |126 unique category code   |
| brand       | Downcased string of brand name. Can be missed.                                                                                                                 | object    | 3444 unique brand   |
| price     | Float price of a product. Present.                                                                                                                               | float64    |  min spend = $0.00, spend = $2547 and average spending = $290   |
| user_id| Permanent user ID.                                                                                                                                                  | int64    | 3022290 user active this month   |
| user_session | Product's category ID                                                                                                                                         | object    |    |
